stdlib/public/stubs/OptionalBridgingHelper.mm

```diff
@@ -0,0 +1,100 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include <Foundation/Foundation.h>
+#include "swift/Basic/Lazy.h"
+#include "swift/Basic/LLVM.h"
+#include "swift/Runtime/Metadata.h"
+#include "swift/Runtime/Mutex.h"
+#include "swift/Runtime/ObjCBridge.h"
+#include <vector>
+
+using namespace swift;
+
+/// Class of sentinel objects used to represent the `nil` value of nested
+/// optionals.
+@interface _SwiftNull : NSObject {
+@public
+  unsigned depth;
+}
+@end
+
+@implementation _SwiftNull : NSObject
+
+- (NSString*)description {
+  return [NSString stringWithFormat:@"<_SwiftNull %u>", self->depth];
+}
+
+- (void)dealloc {
+  [super dealloc];
+}
+
+@end
+
+namespace {
+
+struct SwiftNullSentinelCache {
+  std::vector<id> Cache;
+  StaticReadWriteLock Lock;
+};
+
+static Lazy<SwiftNullSentinelCache> Sentinels;
+
+static id getSentinelForDepth(unsigned depth) {
+  // For unnested optionals, use NSNull.
+  if (depth == 1)
+    return SWIFT_LAZY_CONSTANT([NSNull null]);
```

Comment:
I am fairly certain we have a full obligation from Foundation to maintain the return value as implemented:

```objc
+ (NSNull *)null {
    return (id)kCFNull;
}
```

So the dispatch_once here is not really needed at all.